### PR TITLE
Move clone() from hashtree_test.go into hashtree.go

### PR DESCRIPTION
**Pull request by msteffen**
Use the new clone() in the implementation of HashTree.Merge, and update hashtree_test.go

### CODE REVIEW

File: src/server/pkg/hashtree/hashtree.go [hunk 1]
```diff
@@ -56,6 +56,22 @@ type hashtree struct {
...
+	bts, err := h.Serialize()
+	if err != nil {
+		return nil, errorf(Internal, "could not clone HashTree: %s", err)
+	}
+	h2, err := Deserialize(bts)
+	if err != nil {
+		return nil, errorf(Internal, "could not clone HashTree: "+err.Error())
+	}
+	if h3, ok := h2.(*hashtree); ok {
+		return h3, nil
```
Comment (derekchiang) [inline src/server/pkg/hashtree/hashtree.go]: Not that important but I think this is kind of an anti-pattern.  You should check for `!ok` and return the error in the if clause, and return `h3` in the main function body.

File: src/server/pkg/hashtree/hashtree.go [hunk 2]
```diff
@@ -525,22 +542,12 @@ func (h *hashtree) mergeNode(path string, srcs []HashTree) error {
...
-			return errorf(Internal, "could not deserialize hashtree (due to \"%s\") "+
-				"after merge error: %s", deserializeErr, err)
-		}
-		if htTmp, ok := htInterfaceTmp.(*hashtree); ok {
-			*h = *htTmp
-		} else {
-			return errorf(Internal, "could not convert deserialized hash tree after "+
-				"merge error: %s", err)
-		}
+		*h = *hcopy
```
Comment (derekchiang) [inline src/server/pkg/hashtree/hashtree.go]: This also seems a little strange.  It seems to make more sense to perform the merge on `hcopy`, and only set `*h = *hcopy` if there's no error, as opposed to "resetting" `*h` if there's an error.

File: src/server/pkg/hashtree/hashtree_test.go [hunk 3]
```diff
@@ -75,8 +63,8 @@ func equals(ltmp, rtmp HashTree) bool {
 
 // requireSame compares 'h' to another hash tree (e.g. to make sure that it
 // hasn't changed)
-func requireSame(t *testing.T, ltmp, rtmp HashTree) {
-	l, r := ltmp.(*hashtree), rtmp.(*hashtree)
+func requireSame(t *testing.T, lTmp, rTmp HashTree) {
+	l, r := lTmp.(*hashtree), rTmp.(*hashtree)
```
Comment (derekchiang) [inline src/server/pkg/hashtree/hashtree_test.go]: I think you can use `DeepEqual` directly on interfaces; no need to convert them to concrete types.

File: src/server/pkg/hashtree/hashtree.go (hunk 1 above)
Comment (msteffen) [inline src/server/pkg/hashtree/hashtree.go]: Fixed, it now returns `err` inside the if block

File: src/server/pkg/hashtree/hashtree.go (hunk 2 above)
Comment (msteffen) [inline src/server/pkg/hashtree/hashtree.go]: Fixed this too.

File: src/server/pkg/hashtree/hashtree_test.go (hunk 3 above)
Comment (msteffen) [inline src/server/pkg/hashtree/hashtree_test.go]: Fixed as well

**Pull request merged**